In [1]:
# Load the libararies
import requests
import json
import pandas as pd

## Step 1: Get all the foods

Using the API in the [README](../README.md), get a list of all the foods

In [2]:
url = 'https://fast-retreat-88305.herokuapp.com/foods'


In [3]:
foods = requests.get(url).json()

In [4]:
foods

[{'calories': 370,
  'carbs': 91,
  'category_ids': [1],
  'created_at': '2017-04-17T19:43:29.627Z',
  'fat': 0,
  'id': 26,
  'name': 'Blue Raspberry Slushee',
  'restaurant_id': 1,
  'updated_at': '2017-04-17T19:43:29.627Z',
  'url': 'https://fast-retreat-88305.herokuapp.com/foods/26'},
 {'calories': 570,
  'carbs': 142,
  'category_ids': [1],
  'created_at': '2017-04-17T19:43:29.634Z',
  'fat': 0,
  'id': 27,
  'name': 'Blue Raspberry Slushee',
  'restaurant_id': 1,
  'updated_at': '2017-04-17T19:43:29.634Z',
  'url': 'https://fast-retreat-88305.herokuapp.com/foods/27'},
 {'calories': 740,
  'carbs': 182,
  'category_ids': [1],
  'created_at': '2017-04-17T19:43:29.641Z',
  'fat': 0,
  'id': 28,
  'name': 'Blue Raspberry Slushee',
  'restaurant_id': 1,
  'updated_at': '2017-04-17T19:43:29.641Z',
  'url': 'https://fast-retreat-88305.herokuapp.com/foods/28'},
 {'calories': 480,
  'carbs': 62,
  'category_ids': [3],
  'created_at': '2017-04-17T19:43:29.653Z',
  'fat': 27,
  'id': 29,
  

## Step 2: Get restaurant info for each food

1. For each food dictionary in your foods list, use the `restaurant_id` to get the restaurant information from the API.
2. Add the name from said restaurant to each food dictionary
3. **NOTE:** Cache each restaurant so that you only hit the API once per restaurant. In other words, you should only hit https://fast-retreat-88305.herokuapp.com/restaurants/1 once, even though multiple food items are associated with this particular restaurant.

### Before:
```python
foods = [
    {'calories': 0,
     'carbs': 0,
     'category_id': 1,
     'created_at': '2017-04-17T19:43:29.444Z',
     'fat': 0,
     'id': 1,
     'name': 'A&W® Diet Root Beer',
     'restaurant_id': 1,
     'updated_at': '2017-04-17T19:43:29.444Z',
     'url': 'https://fast-retreat-88305.herokuapp.com/foods/1'},
     ...
]
```

### After:

Note the addition of **`'name': 'A&W® Diet Root Beer'`** to the dictionary.

```python
foods = [
    {'calories': 0,
     'carbs': 0,
     'category_id': 1,
     'created_at': '2017-04-17T19:43:29.444Z',
     'fat': 0,
     'id': 1,
     'name': 'A&W® Diet Root Beer',
     'restaurant': 'A&W Restaurants',
     'restaurant_id': 1,
     'updated_at': '2017-04-17T19:43:29.444Z',
     'url': 'https://fast-retreat-88305.herokuapp.com/foods/1'},
     ...
]
```

In [5]:
rest_url = 'https://fast-retreat-88305.herokuapp.com/restaurants'
rest_info = requests.get(rest_url).json()

In [6]:
rest_info[0]

{'created_at': '2017-04-17T19:43:26.708Z',
 'id': 1,
 'name': 'A&W Restaurants',
 'updated_at': '2017-04-17T19:43:26.708Z',
 'url': 'https://fast-retreat-88305.herokuapp.com/restaurants/1'}

** Code for category data **

In [7]:
cat_url = 'https://fast-retreat-88305.herokuapp.com/categories'
cat_info = requests.get(cat_url).json()

In [8]:
cat_info[0]

{'created_at': '2017-04-17T19:43:29.423Z',
 'id': 1,
 'name': 'Drinks',
 'updated_at': '2017-04-17T20:05:35.796Z',
 'url': 'https://fast-retreat-88305.herokuapp.com/categories/1'}

In [9]:
foods[0]

{'calories': 370,
 'carbs': 91,
 'category_ids': [1],
 'created_at': '2017-04-17T19:43:29.627Z',
 'fat': 0,
 'id': 26,
 'name': 'Blue Raspberry Slushee',
 'restaurant_id': 1,
 'updated_at': '2017-04-17T19:43:29.627Z',
 'url': 'https://fast-retreat-88305.herokuapp.com/foods/26'}

In [10]:
for food in foods:
    for rest in rest_info:
        if food['restaurant_id'] == rest['id']:
            food['restaurant'] = rest['name']
# adding restaurant name

In [11]:
foods[0]

{'calories': 370,
 'carbs': 91,
 'category_ids': [1],
 'created_at': '2017-04-17T19:43:29.627Z',
 'fat': 0,
 'id': 26,
 'name': 'Blue Raspberry Slushee',
 'restaurant': 'A&W Restaurants',
 'restaurant_id': 1,
 'updated_at': '2017-04-17T19:43:29.627Z',
 'url': 'https://fast-retreat-88305.herokuapp.com/foods/26'}

In [12]:
for food in foods:    
    for cat in cat_info:
        if food['restaurant_id'] == cat['id']:
            food['category'] = cat['name']
# adding category information

In [13]:
foods[0]

{'calories': 370,
 'carbs': 91,
 'category': 'Drinks',
 'category_ids': [1],
 'created_at': '2017-04-17T19:43:29.627Z',
 'fat': 0,
 'id': 26,
 'name': 'Blue Raspberry Slushee',
 'restaurant': 'A&W Restaurants',
 'restaurant_id': 1,
 'updated_at': '2017-04-17T19:43:29.627Z',
 'url': 'https://fast-retreat-88305.herokuapp.com/foods/26'}

In [14]:
foods_df = pd.DataFrame(foods)

In [15]:
foods_df.shape

(4976, 12)

## Step 2: Get category info for each food

1. For each food dictionary in your foods list, use the `category_id` to get the category information from the API.
2. Add the name from said category to each food dictionary
3. **NOTE:** Cache each category so that you only hit the API once per category. In other words, you should only hit https://fast-retreat-88305.herokuapp.com/categories/1 once, even though multiple food items are associated with this particular category.

### Before:
```python
foods = [
    {'calories': 0,
     'carbs': 0,
     'category_id': 1,
     'created_at': '2017-04-17T19:43:29.444Z',
     'fat': 0,
     'id': 1,
     'name': 'A&W® Diet Root Beer',
     'restaurant': 'A&W Restaurants',
     'restaurant_id': 1,
     'updated_at': '2017-04-17T19:43:29.444Z',
     'url': 'https://fast-retreat-88305.herokuapp.com/foods/1'},
     ...
]
```

### After:

Note the addition of **`'category': 'Drinks'`** to the dictionary.

```python
foods = [
    {'calories': 0,
     'carbs': 0,
     'category': 'Drinks',
     'category_id': 1,
     'created_at': '2017-04-17T19:43:29.444Z',
     'fat': 0,
     'id': 1,
     'name': 'A&W® Diet Root Beer',
     'restaurant': 'A&W Restaurants',
     'restaurant_id': 1,
     'updated_at': '2017-04-17T19:43:29.444Z',
     'url': 'https://fast-retreat-88305.herokuapp.com/foods/1'},
     ...
]
```

In [16]:
foods_df.head()

calories  carbs category category_ids                created_at   fat  id  \
0     370.0   91.0   Drinks          [1]  2017-04-17T19:43:29.627Z   0.0  26   
1     570.0  142.0   Drinks          [1]  2017-04-17T19:43:29.634Z   0.0  27   
2     740.0  182.0   Drinks          [1]  2017-04-17T19:43:29.641Z   0.0  28   
3     480.0   62.0   Drinks          [3]  2017-04-17T19:43:29.653Z  27.0  29   
4     350.0   45.0   Drinks          [3]  2017-04-17T19:43:29.659Z  16.0  30   

                            name       restaurant  restaurant_id  \
0         Blue Raspberry Slushee  A&W Restaurants            1.0   
1         Blue Raspberry Slushee  A&W Restaurants            1.0   
2         Blue Raspberry Slushee  A&W Restaurants            1.0   
3    Breaded Onion Rings (Large)  A&W Restaurants            1.0   
4  Breaded Onion Rings (Regular)  A&W Restaurants            1.0   

                 updated_at                                                url  
0  2017-04-17T19:43:29.627Z  https://fast-retreat-88305.herokuapp.com/foods/26  
1  2017-04-17T19:43:29.634Z  https://fast-retreat-88305.herokuapp.com/foods/27  
2  2017-04-17T19:43:29.641Z  https://fast-retreat-88305.herokuapp.com/foods/28  
3  2017-04-17T19:43:29.653Z  https://fast-retreat-88305.herokuapp.com/foods/29  
4  2017-04-17T19:43:29.659Z  https://fast-retreat-88305.herokuapp.com/foods/30

In [17]:
new_foods = foods_df.drop(['category_ids', 'created_at', 'restaurant_id', 'updated_at', 'url'], axis=1)
new_foods.head()

calories  carbs category   fat  id                           name  \
0     370.0   91.0   Drinks   0.0  26         Blue Raspberry Slushee   
1     570.0  142.0   Drinks   0.0  27         Blue Raspberry Slushee   
2     740.0  182.0   Drinks   0.0  28         Blue Raspberry Slushee   
3     480.0   62.0   Drinks  27.0  29    Breaded Onion Rings (Large)   
4     350.0   45.0   Drinks  16.0  30  Breaded Onion Rings (Regular)   

        restaurant  
0  A&W Restaurants  
1  A&W Restaurants  
2  A&W Restaurants  
3  A&W Restaurants  
4  A&W Restaurants

## Step 3: Create a `pandas` DataFrame from your list of foods

**NOTE:** The shape of your DataFrame should be 4976 x 12

## Step 4: Clean up the DataFrame

1. Drop any superfluous columns
2. Capitalize each column name **in one line of code**

In [18]:
new_foods.columns  = [x.title() for x in new_foods.columns]

In [19]:
new_foods.head()

Calories  Carbs Category   Fat  Id                           Name  \
0     370.0   91.0   Drinks   0.0  26         Blue Raspberry Slushee   
1     570.0  142.0   Drinks   0.0  27         Blue Raspberry Slushee   
2     740.0  182.0   Drinks   0.0  28         Blue Raspberry Slushee   
3     480.0   62.0   Drinks  27.0  29    Breaded Onion Rings (Large)   
4     350.0   45.0   Drinks  16.0  30  Breaded Onion Rings (Regular)   

        Restaurant  
0  A&W Restaurants  
1  A&W Restaurants  
2  A&W Restaurants  
3  A&W Restaurants  
4  A&W Restaurants

In [20]:
new_foods = new_foods.reindex(columns=['Restaurant', 'Name', 'Id', 'Category', 'Calories', 'Carbs', 'Fat'])

In [22]:
new_foods.isnull().sum()

Restaurant       1
Name             1
Id               0
Category      1435
Calories         2
Carbs          119
Fat              1
dtype: int64

In [24]:
new_foods.fillna(value='N/A', inplace=True)

In [25]:
new_foods.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4976 entries, 0 to 4975
Data columns (total 7 columns):
Restaurant    4976 non-null object
Name          4976 non-null object
Id            4976 non-null int64
Category      4976 non-null object
Calories      4976 non-null object
Carbs         4976 non-null object
Fat           4976 non-null object
dtypes: int64(1), object(6)
memory usage: 272.2+ KB
